# AAI614: Data Science & its Applications

*Notebook 2.6: Practice with Data Collections*

<a href="https://colab.research.google.com/github/harmanani/AAI614/blob/main/Week%202/Notebook2.6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import pandas as pd
from pandas import Series, DataFrame

import ssl

ssl._create_default_https_context = ssl._create_unverified_context

In [2]:
# Import the taxi info from both January and July 2019.
# Include the following columns: tpep_pickup_datetime, passenger_count, trip_distance,
# fare_amount, extra, mta_tax, tip_amount, tolls_amount, improvement_surcharge,
# total_amount, and congestion_surcharge.

filenames = ['https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-01aa.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-01aq.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-01bg.csv',
             'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-07aa.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-01ab.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-01ar.csv',
            'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-07ab.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-07aq.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-01ac.csv',
            'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-01as.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-07ab.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-07ar.csv',
            'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-01ad.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-01at.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-07ac.csv',
            'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-07as.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-01ae.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-01au.csv',
            'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-07ad.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-07at.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-01af.csv',
            'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-01av.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-07ae.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-07au.csv',
            'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-01ag.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-01aw.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-07af.csv',
            'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-07av.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-01ah.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-01ax.csv',
            'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-07ag.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-07aw.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-01ai.csv', 
            'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-01ay.csv',
            'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-07ah.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-07ax.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-01aj.csv',
            'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-01az.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-07ai.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-07ay.csv',
            'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-01ak.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-01ba.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-07aj.csv',
            'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-07az.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-01al.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-01bb.csv', 
            'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-07ak.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-07ba.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-01am.csv',
            'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-01bc.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-07al.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-07bb.csv',
            'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-01an.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-01bd.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-07am.csv',
            'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-01ao.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-01be.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-07an.csv',
            'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-01ap.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-01bf.csv', 'https://raw.githubusercontent.com/harmanani/AAI614/refs/heads/main/Week%202/nyc_taxi/nyc_taxi_2019-07ao.csv']

headers = ['VendorID', 'tpep_pickup_datetime', 'tpep_dropoff_datetime', 'passenger_count', 'trip_distance', 'RatecodeID', 'store_and_fwd_flag', 'PULocationID', 'DOLocationID', 'payment_type', 'fare_amount', 'extra', 'mta_tax', 'tip_amount', 'tolls_amount', 'improvement_surcharge', 'total_amount', 'congestion_surcharge']
cols = ['tpep_pickup_datetime', 'passenger_count', 'trip_distance', 'fare_amount', 'extra', 'mta_tax', 'tip_amount', 'tolls_amount', 'improvement_surcharge', 'total_amount', 'congestion_surcharge']

all_dfs = [pd.read_csv(one_filename, 
                       names = headers, 
                       header = 1, 
                       low_memory=False,
                       usecols = cols, 
                       parse_dates = ['tpep_pickup_datetime']) 
                       for one_filename in filenames]

df = pd.concat(all_dfs)

df.head()

,tpep_pickup_datetime,passenger_count,trip_distance,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge
0,2019-01-01 00:59:47,1.0,2.6,14.0,0.5,0.5,1.0,0.00,0.3,16.30,NaN
1,2018-12-21 13:48:30,3.0,0.0,4.5,0.5,0.5,0.0,0.00,0.3,5.80,NaN
2,2018-11-28 15:52:25,5.0,0.0,3.5,0.5,0.5,0.0,0.00,0.3,7.55,NaN
3,2018-11-28 15:56:57,5.0,0.0,52.0,0.0,0.5,0.0,0.00,0.3,55.55,NaN
4,2018-11-28 16:25:49,5.0,0.0,3.5,0.5,0.5,0.0,5.76,0.3,13.31,NaN


In [3]:
# Create a new column, pre_tip_amount
df['pre_tip_amount'] = (
    df
    [['fare_amount', 'extra', 'mta_tax', 'tolls_amount', 
      'improvement_surcharge', 'congestion_surcharge']]
    .sum(axis='columns')
)

In [4]:
# Create a new column, tip_percentage, showing the percent of the fare_amount that people tipped
df['tip_percentage'] = df['tip_amount'] / df['pre_tip_amount']

In [5]:
# What was the overall tip percentage?
df['tip_percentage'].mean()

np.float64(0.12993117643752727)

In [6]:
# How many times did people tip more than 100%?
(df['tip_percentage'] > 1).value_counts()

tip_percentage
False    13971177
True         7806
Name: count, dtype: int64

In [7]:
# What percent of taxi riders give no tip at all?
(df['tip_percentage'] == 0).value_counts(normalize=True)

tip_percentage
False    0.678608
True     0.321392
Name: proportion, dtype: float64

In [8]:
# On which day of the week do people tip the greatest percentage?
# Mon = 0, Tues = 1, Wed = 2, Thu = 3, Fri = 4, Sat = 5, Sun = 6
df['tpep_pickup_datetime'] = pd.to_datetime(df['tpep_pickup_datetime'], errors = 'coerce')
df.groupby(df['tpep_pickup_datetime'].dt.day_of_week)['tip_percentage'].mean().sort_values(ascending=False)

tpep_pickup_datetime
3.0    0.134325
2.0    0.131881
1.0    0.130681
4.0    0.129410
0.0    0.128723
6.0    0.126634
5.0    0.125801
Name: tip_percentage, dtype: float64

In [9]:
# At which hour do people tip the greatest percentage?
df.groupby(df['tpep_pickup_datetime'].dt.hour)['tip_percentage'].mean().sort_values(ascending=False)

tpep_pickup_datetime
22.0    0.138875
20.0    0.138077
21.0    0.137589
8.0     0.137015
19.0    0.135120
23.0    0.134812
18.0    0.133217
9.0     0.132800
7.0     0.131954
0.0     0.131289
1.0     0.130993
2.0     0.130831
17.0    0.128518
10.0    0.127014
11.0    0.124796
16.0    0.124588
13.0    0.124417
12.0    0.124323
14.0    0.123623
15.0    0.123450
3.0     0.121132
6.0     0.119908
4.0     0.118884
5.0     0.112030
Name: tip_percentage, dtype: float64

In [10]:
# Do people tip more, on average, in January or July?
df.groupby(df['tpep_pickup_datetime'].dt.month)['tip_percentage'].mean().sort_values(ascending=False)

tpep_pickup_datetime
5.0     0.200000
8.0     0.158099
3.0     0.148046
9.0     0.141431
1.0     0.137011
2.0     0.132224
7.0     0.121331
12.0    0.109367
6.0     0.108901
10.0    0.100000
4.0     0.074877
11.0    0.046026
Name: tip_percentage, dtype: float64

In [11]:
# What was the 1-day period in our data set when people tipped the greatest percentage?
df = df.set_index('tpep_pickup_datetime')
df.resample('1D')['tip_percentage'].mean().sort_values(ascending=False).head(10)

tpep_pickup_datetime
2019-02-13    0.358127
2019-02-25    0.250000
2019-08-20    0.241865
2019-05-20    0.200000
2019-08-15    0.200000
2019-09-22    0.200000
2019-11-27    0.200000
2019-08-10    0.200000
2019-09-24    0.200000
2010-08-08    0.200000
Name: tip_percentage, dtype: float64

In [12]:
# Try again, within our range
df = df.sort_index()

df = pd.concat([
    df[(df.index >= '2019-01-01') & (df.index < '2019-02-01')],
    df[(df.index >= '2019-07-01') & (df.index < '2019-08-01')]
])

df.resample('1D')['tip_percentage'].mean().sort_values(ascending=False).head(10)

tpep_pickup_datetime
2019-01-31    0.144348
2019-01-30    0.143530
2019-01-24    0.143434
2019-01-22    0.142769
2019-01-15    0.142330
2019-01-29    0.141331
2019-01-10    0.141291
2019-01-16    0.141147
2019-01-17    0.140356
2019-01-23    0.140309
Name: tip_percentage, dtype: float64

In [13]:
df.resample('1ME')['tip_percentage'].mean().dropna()

tpep_pickup_datetime
2019-01-31    0.137012
2019-07-31    0.121331
Name: tip_percentage, dtype: float64

# Questions

1. We saw that 32 percent of riders don't tip at all. Of those who *do*, what percentage do they tip, on average?

In [14]:
# Of the riders who left a tip, what percentage did they tip on average?
df[df['tip_percentage'] > 0]['tip_percentage'].mean()

np.float64(0.19148866955267874)

Riders who leave a tip give about 19.1% on average.

2. How many of the rides in our data set, supposedly from January and July 2019, are from outside of those dates?

In [15]:
# Check the original (unfiltered) pickup dates to see how many rides fall outside of January and July 2019
original_dates = pd.to_datetime(pd.concat(all_dfs)['tpep_pickup_datetime'], errors='coerce')

in_range = (
    ((original_dates.dt.year == 2019) & (original_dates.dt.month == 1)) |
    ((original_dates.dt.year == 2019) & (original_dates.dt.month == 7))
)

(~in_range).sum()

np.int64(823)

823* rides have pickup dates outside of January and July 2019, some are from other months of 2019, and some are from entirely different years.

3. Looking only at dates in January and July, on what week did passengers tip the greatest percentage?

In [16]:
# Looking only at January and July, which week had the highest average tip percentage?
df.resample('1W')['tip_percentage'].mean().sort_values(ascending=False).head(10)

tpep_pickup_datetime
2019-02-03    0.141978
2019-01-27    0.138930
2019-01-20    0.138536
2019-01-13    0.137901
2019-01-06    0.126983
2019-08-04    0.124910
2019-07-14    0.123459
2019-07-28    0.123036
2019-07-21    0.122778
2019-07-07    0.114103
Name: tip_percentage, dtype: float64

The week labeled 2019-02-03 had the highest average tip percentage at about 14.2%.